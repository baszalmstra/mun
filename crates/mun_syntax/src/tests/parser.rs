use crate::SourceFile;

fn snapshot_test(text: &str) {
    let text = text.trim().replace("\n    ", "\n");
    let file = SourceFile::parse(&text);
    insta::assert_snapshot!(insta::_macro_support::AutoName, file.debug_dump(), &text);
}

#[test]
fn empty() {
    snapshot_test(r#""#);
}

#[test]
fn function() {
    snapshot_test(
        r#"
    // Source file comment

    // Comment that belongs to the function
    fn a() {}
    fn b(value:number) {}
    export fn c():never {}
    fn b(value:number):number {}"#,
    );
}

#[test]
fn block() {
    snapshot_test(
        r#"
    fn foo() {
        let a;
        let b:i32;
        let c:string;
    }"#,
    );
}

#[test]
fn literals() {
    snapshot_test(
        r#"
    fn foo() {
        let a = true;
        let b = false;
        let c = 1;
        let d = 1.12;
        let e = "Hello, world!"
    }
    "#,
    );
}

#[test]
fn unary_expr() {
    snapshot_test(
        r#"
    fn foo() {
        let a = --3;
        let b = !!true;
    }
    "#,
    )
}

#[test]
fn binary_expr() {
    snapshot_test(
        r#"
    fn foo() {
        let a = 3+4*5
        let b = 3*4+10/2
    }
    "#,
    )
}

#[test]
fn expression_statement() {
    snapshot_test(
        r#"
    fn foo() {
        let a = "hello"
        let b = "world"
        let c
        b = "Hello, world!"
        !-5+2*(a+b);
        -3
    }
    "#,
    )
}

#[test]
fn function_calls() {
    snapshot_test(
        r#"
    fn bar(i:number) { }
    fn foo(i:number) {
      bar(i+1)
    }
    "#,
    )
}

#[test]
fn patterns() {
    snapshot_test(
        r#"
    fn main(_:number) {
       let a = 0;
       let _ = a;
    }
    "#,
    )
}

#[test]
fn compare_operands() {
    snapshot_test(
        r#"
    fn main() {
        let _ = a==b;
        let _ = a == b;
        let _ = a != b;
        let _ = a < b;
        let _ = a > b;
        let _ = a <= b;
        let _ = a >= b;
    }
    "#,
    )
}

#[test]
fn if_expr() {
    snapshot_test(
        r#"
    fn bar() {
        if true {};
        if true {} else {};
        if true {} else if false {} else {};
        if {true} {} else {}
    }
    "#,
    );
}

#[test]
fn block_expr() {
    snapshot_test(
        r#"
    fn bar() {
        {3}
    }
    "#,
    );
}

#[test]
fn return_expr() {
    snapshot_test(
        r#"
    fn foo() {
        return;
        return 50;
    }
    "#,
    )
}

#[test]
fn loop_expr() {
    snapshot_test(
        r#"
    fn foo() {
        loop {}
    }"#,
    )
}

#[test]
fn break_expr() {
    snapshot_test(
        r#"
    fn foo() {
        break;
        if break { 3; }
        if break 4 { 3; }
    }
    "#,
    )
}

#[test]
fn while_expr() {
    snapshot_test(
        r#"
    fn foo() {
        while true {};
        while { true } {};
    }
    "#,
    )
}

#[test]
fn range_expr() {
    snapshot_test(
        r#"
    fn foo() {
        1..2;
        3..;
        ..4;
        ..;
        5..=6;
        ..=7;
        3..=; // Error: an inclusive range must have an end expression
        ..=;  // Error: an inclusive range must have an end expression
    }
    "#,
    )
}

#[test]
fn postfix_range() {
    snapshot_test(
        r#"
    fn foo() {
        let x = 1..;
        if 1.. {};
    }
    "#,
    )
}
