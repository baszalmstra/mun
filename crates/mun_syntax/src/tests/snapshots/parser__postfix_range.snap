---
source: crates/mun_syntax/src/tests/parser.rs
expression: "fn foo() {\n    let x = 1..;\n    if 1.. {};\n}"
---
SOURCE_FILE@[0; 44)
  FUNCTION_DEF@[0; 44)
    FN_KW@[0; 2) "fn"
    WHITESPACE@[2; 3) " "
    NAME@[3; 6)
      IDENT@[3; 6) "foo"
    PARAM_LIST@[6; 8)
      L_PAREN@[6; 7) "("
      R_PAREN@[7; 8) ")"
    WHITESPACE@[8; 9) " "
    BLOCK_EXPR@[9; 44)
      L_CURLY@[9; 10) "{"
      WHITESPACE@[10; 15) "\n    "
      LET_STMT@[15; 27)
        LET_KW@[15; 18) "let"
        WHITESPACE@[18; 19) " "
        BIND_PAT@[19; 20)
          NAME@[19; 20)
            IDENT@[19; 20) "x"
        WHITESPACE@[20; 21) " "
        EQ@[21; 22) "="
        WHITESPACE@[22; 23) " "
        RANGE_EXPR@[23; 26)
          LITERAL@[23; 24)
            INT_NUMBER@[23; 24) "1"
          DOTDOT@[24; 26) ".."
        SEMI@[26; 27) ";"
      WHITESPACE@[27; 32) "\n    "
      EXPR_STMT@[32; 42)
        IF_EXPR@[32; 41)
          IF_KW@[32; 34) "if"
          WHITESPACE@[34; 35) " "
          CONDITION@[35; 38)
            RANGE_EXPR@[35; 38)
              LITERAL@[35; 36)
                INT_NUMBER@[35; 36) "1"
              DOTDOT@[36; 38) ".."
          WHITESPACE@[38; 39) " "
          BLOCK_EXPR@[39; 41)
            L_CURLY@[39; 40) "{"
            R_CURLY@[40; 41) "}"
        SEMI@[41; 42) ";"
      WHITESPACE@[42; 43) "\n"
      R_CURLY@[43; 44) "}"

